File: Src/java/engine/src/test/java/org/opencds/cqf/cql/engine/execution/ProfilingTest.java
package org.opencds.cqf.cql.engine.execution;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;

import java.io.IOException;
import java.io.StringWriter;
import java.util.Map;
import java.util.Set;
import org.junit.jupiter.api.Test;

public class ProfilingTest extends CqlTestBase {
    private Profile produceProfile() {
        final var environment = new Environment(getLibraryManager());
        final var engine = new CqlEngine(
                environment,
                Set.of(
                        CqlEngine.Options.EnableExpressionCaching,
                        CqlEngine.Options.EnableProfiling)); // TODO: engine options to CqlTestBase.getEngine instead?
        final var result = engine.evaluate("ProfilingTest");
        final var debugResult = result.getDebugResult();
        return debugResult.getProfile();
    }

    @Test
    public void profilingSmoke() {
        final var profile = produceProfile();
        final var tree = profile.getTree();
        assertThat(tree.expression, nullValue());
        assertThat(
                tree,
                new NodeMatcher(
                        nullValue(String.class),
                        equalTo(1L),
                        greaterThanOrEqualTo(0L),
                        equalTo(1L),
                        Map.of(
                                "Unfiltered",
                                Map.of(
                                        "E1",
                                        new NodeMatcher(
                                                equalTo("Unfiltered"),
                                                equalTo(1L),
                                                greaterThanOrEqualTo(0L),
                                                equalTo(1L),
                                                Map.of(
                                                        "Unfiltered",
                                                        Map.of(
                                                                "G",
                                                                new NodeMatcher(
                                                                        equalTo("Unfiltered"),
                                                                        equalTo(1L),
                                                                        greaterThanOrEqualTo(0L),
                                                                        equalTo(1L),
                                                                        Map.of(
                                                                                "Unfiltered",
                                                                                Map.of(
                                                                                        "F",
                                                                                        new NodeMatcher(
                                                                                                equalTo("Unfiltered"),
                                                                                                equalTo(12L),
                                                                                                greaterThanOrEqualTo(
                                                                                                        0L),
                                                                                                equalTo(12L)))))))),
                                        "E2",
                                        new NodeMatcher(
                                                equalTo("Unfiltered"),
                                                equalTo(1L),
                                                greaterThanOrEqualTo(0L),
                                                equalTo(1L),
                                                Map.of(
                                                        "Unfiltered",
                                                        Map.of(
                                                                "G",
                                                                new NodeMatcher(
                                                                        equalTo("Unfiltered"),
                                                                        equalTo(1L),
                                                                        greaterThanOrEqualTo(0L),
                                                                        equalTo(1L),
                                                                        Map.of(
                                                                                "Unfiltered",
                                                                                Map.of(
                                                                                        "F",
                                                                                        new NodeMatcher(
                                                                                                equalTo("Unfiltered"),
                                                                                                equalTo(22L),
                                                                                                greaterThanOrEqualTo(
                                                                                                        0L),
                                                                                                equalTo(22L)))))))),
                                        "E3",
                                        new NodeMatcher(
                                                equalTo("Unfiltered"),
                                                equalTo(1L),
                                                greaterThanOrEqualTo(0L),
                                                equalTo(1L),
                                                Map.of(
                                                        "Unfiltered",
                                                        Map.of(
                                                                "E2",
                                                                new NodeMatcher(
                                                                        equalTo("Unfiltered"),
                                                                        equalTo(1L),
                                                                        greaterThanOrEqualTo(0L),
                                                                        equalTo(0L))))),
                                        "E4",
                                        new NodeMatcher(
                                                equalTo("Unfiltered"),
                                                equalTo(1L),
                                                greaterThanOrEqualTo(0L),
                                                equalTo(1L),
                                                Map.of(
                                                        "Unfiltered",
                                                        Map.of(
                                                                "E2",
                                                                new NodeMatcher(
                                                                        equalTo("Unfiltered"),
                                                                        equalTo(1L),
                                                                        greaterThanOrEqualTo(0L),
                                                                        equalTo(0L)))))))));
    }

    @Test
    public void renderSmoke() throws IOException {
        final var profile = produceProfile();
        final var writer = new StringWriter();
        profile.render(writer);
        writer.close();
        final var svgString = writer.toString();
        assertThat(svgString, startsWith("<svg>"));
        assertThat(svgString, endsWith("</svg>\n"));
    }
}

File: Src/java/engine/src/jvmTest/kotlin/org/opencds/cqf/cql/engine/execution/ProfilingTest.kt
package org.opencds.cqf.cql.engine.execution

import java.io.IOException
import kotlinx.io.Buffer
import kotlinx.io.readString
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers
import org.hamcrest.Matchers.equalTo
import org.hamcrest.Matchers.greaterThanOrEqualTo
import org.hamcrest.Matchers.nullValue
import org.junit.jupiter.api.Test

class ProfilingTest : CqlTestBase() {
    private fun produceProfile(): Profile {
        val environment = Environment(libraryManager)
        val engine =
            CqlEngine(
                environment,
                mutableSetOf(
                    CqlEngine.Options.EnableExpressionCaching,
                    CqlEngine.Options.EnableProfiling,
                ),
            ) // TODO: engine options to CqlTestBase.getEngine instead?
        val result = engine.evaluate { library("ProfilingTest") }.onlyResultOrThrow
        val debugResult = result.debugResult
        return debugResult!!.profile!!
    }

    @Test
    fun profilingSmoke() {
        val profile = produceProfile()
        val tree = profile.tree
        assertThat(tree.expression, nullValue())

        assertThat(
            tree,
            NodeMatcher(
                nullValue(String::class.java),
                equalTo(1L),
                greaterThanOrEqualTo(0L),
                equalTo(1L),
                mapOf(
                    "Unfiltered" to
                        mapOf(
                            "E1" to
                                NodeMatcher(
                                    equalTo("Unfiltered"),
                                    equalTo(1L),
                                    greaterThanOrEqualTo(0L),
                                    equalTo(1L),
                                    mapOf(
                                        "Unfiltered" to
                                            mapOf(
                                                "G" to
                                                    NodeMatcher(
                                                        equalTo("Unfiltered"),
                                                        equalTo(1L),
                                                        greaterThanOrEqualTo(0L),
                                                        equalTo(1L),
                                                        mapOf(
                                                            "Unfiltered" to
                                                                mapOf(
                                                                    "F" to
                                                                        NodeMatcher(
                                                                            equalTo("Unfiltered"),
                                                                            equalTo(12L),
                                                                            greaterThanOrEqualTo(
                                                                                0L
                                                                            ),
                                                                            equalTo(12L),
                                                                        )
                                                                )
                                                        ),
                                                    )
                                            )
                                    ),
                                ),
                            "E2" to
                                NodeMatcher(
                                    equalTo("Unfiltered"),
                                    equalTo(1L),
                                    greaterThanOrEqualTo(0L),
                                    equalTo(1L),
                                    mapOf(
                                        "Unfiltered" to
                                            mapOf(
                                                "G" to
                                                    NodeMatcher(
                                                        equalTo("Unfiltered"),
                                                        equalTo(1L),
                                                        greaterThanOrEqualTo(0L),
                                                        equalTo(1L),
                                                        mapOf(
                                                            "Unfiltered" to
                                                                mapOf(
                                                                    "F" to
                                                                        NodeMatcher(
                                                                            equalTo("Unfiltered"),
                                                                            equalTo(22L),
                                                                            greaterThanOrEqualTo(
                                                                                0L
                                                                            ),
                                                                            equalTo(22L),
                                                                        )
                                                                )
                                                        ),
                                                    )
                                            )
                                    ),
                                ),
                            "E3" to
                                NodeMatcher(
                                    equalTo("Unfiltered"),
                                    equalTo(1L),
                                    greaterThanOrEqualTo(0L),
                                    equalTo(1L),
                                    mapOf(
                                        "Unfiltered" to
                                            mapOf(
                                                "E2" to
                                                    NodeMatcher(
                                                        equalTo("Unfiltered"),
                                                        equalTo(1L),
                                                        greaterThanOrEqualTo(0L),
                                                        equalTo(0L),
                                                    )
                                            )
                                    ),
                                ),
                            "E4" to
                                NodeMatcher(
                                    equalTo("Unfiltered"),
                                    equalTo(1L),
                                    greaterThanOrEqualTo(0L),
                                    equalTo(1L),
                                    mapOf(
                                        "Unfiltered" to
                                            mapOf(
                                                "E2" to
                                                    NodeMatcher(
                                                        equalTo("Unfiltered"),
                                                        equalTo(1L),
                                                        greaterThanOrEqualTo(0L),
                                                        equalTo(0L),
                                                    )
                                            )
                                    ),
                                ),
                        )
                ),
            ),
        )
    }

    @Test
    @Throws(IOException::class)
    fun renderSmoke() {
        val profile = produceProfile()
        val buffer = Buffer()
        profile.render(buffer)
        val svgString = buffer.readString()
        assertThat(svgString, Matchers.startsWith("<svg>"))
        assertThat(svgString, Matchers.endsWith("</svg>\n"))
    }
}
